app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,197 @@
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        for key in payload.keys():
+            self.assertEqual(payload[key], getattr(recipe, key))
+
+    def test_create_recipe_with_ingredients(self):
+        """Test creating a recipe with ingredients"""
+        recipe = sample_recipe()
+        ingredient1 = sample_ingredient()
+        ingredient2 = sample_ingredient(name='Tomato')
+        ingredients = [ingredient1, ingredient2]
+        payload = {
+            'name': recipe.name,
+            'description': recipe.description,
+            'ingredients': ingredients,
+        }
+
+        res = self.client.post(RECIPES_URL, payload, format='json')
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        ingredients = recipe.ingredients.all().order_by('id')
+        ingredient_serializer = IngredientSerializer(ingredients, many=True)
+        self.assertEqual(ingredients.count(), 2)
+        self.assertEqual(
+            ingredient1['name'], ingredient_serializer.data[0]['name']
+        )
+        self.assertEqual(
+            ingredient2['name'], ingredient_serializer.data[1]['name']
+        )
+
+    def test_recipe_update_add_ingredient(self):
+        """Test updating a recipe with an ingredient"""
+        recipe = sample_recipe()
+
+        payload = {
+            'name': 'Yummy scrummy chicken',
+            'description': 'Wow this tastes good',
+            'ingredients': [
+                {'name': 'Chicken'}
+            ]
+        }
+
+        url = detail_url(recipe.id)
+        self.client.patch(url, data=payload, format='json')
+        res = self.client.get(url)
+
+        recipe.refresh_from_db()
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(recipe.name, res.data['name'], )
+        self.assertEqual(len(ingredients), 1)
+
+        for ingredient in ingredients:
+            self.assertIn(ingredient.name, 'Chicken')
+
+    def test_recipe_update_add_ingredients(self):
+        """Test updating a recipe with new ingredients"""
+        recipe = sample_recipe()
+        ingredient = Ingredient.objects.create(
+            name='Basil',
+            recipe=recipe
+        )
+        recipe.ingredients.add(ingredient)
+        payload = {
+            'ingredients': [
+                {'name': 'Basil'},
+                {'name': 'Chicken'}
+            ]
+        }
+
+        url = detail_url(recipe.id)
+        self.client.patch(url, data=payload, format='json')
+        res = self.client.get(url)
+
+        recipe.refresh_from_db()
+
+        res_recipe = res.data
+        ingredients = res_recipe['ingredients']
+
+        self.assertEqual(recipe.name, res.data['name'], )
+        self.assertEqual(len(ingredients), 2)
+        for ingredient in ingredients:
+            self.assertIn(ingredient['name'], 'Chicken Basil')
```

Comment:
IMO this is a bit hard to read, matching each full ingredient name to a substring of a single string. For this sort of thing another approach would be to create a set from the 2 ingredient names from `ingredients` (`set(i.name for i in ingredients)`) and compare equal to `{'Chicken', 'Basil'}`